### ISSUE TITLE
Atomic file updates

### ISSUE THREAD

**Issue by tam7t** [kubernetes-sigs/secrets-store-csi-driver#500]
**Describe the solution you'd like**

The implementation of K8s Secrets uses symbolic links to present an atomic update of the entire mount to the application when secrets change (some K8s discussion of [secret updates](https://github.com/kubernetes/kubernetes/issues/18372)).

The [current implementation](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/481) in the driver and most (all?) providers updates a file at a time.

Creating this bug to track any observed issues with current behavior and whether we can switch to using `kubernetes/pkg/volume/util/atomic_writer.go` directly without modification.

**Anything else you would like to add:**

In the CSI Driver the targetPath (root of tmpfs) is of the form:

```
/var/lib/kubelet/pods/<pod id>/volumes/kubernetes.io~csi/<volume id>/mount
```

This path's contents are directly visible to the pod. In order to use the `atomic_writer` it would need to manage files in `../data_tmp` and other parent paths:

```
/var/lib/kubelet/pods/<pod id>/volumes/kubernetes.io~csi/<volume id>/data_tmp
```

Since the kubelet(?) populates `target_path` in `NodePublishVolumeRequest`, I'm unclear whether its possible to write to files in the parent directory. [ref](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodepublishvolume)

**Comment by tam7t** [comment]
Also note that we currently create the tmpfs mount at the `target_path`, to use the `atomic_writer.go` the other sibling folders that atomic_writer uses to stage files should _also_ be tmpfs.

**Comment by tam7t** [comment]
🤦‍♂️ 

I may have missread the atomic_writer docs, it seems that the `..data` paths are not in the parent directory but part of the filenames and visible to the pod itself. From the host:

```
$ sudo ls -la /var/lib/kubelet/pods/903a43b4-4553-4e05-acfa-669e842e1dde/volumes/kubernetes.io~secret/mypodserviceaccount-token-4xmpw
total 4
drwxrwxrwt 3 root root  140 Apr 20 19:24 .
drwxr-xr-x 3 root root 4096 Apr 20 19:24 ..
drwxr-xr-x 2 root root  100 Apr 20 19:24 ..2021_04_20_19_24_41.755804101
lrwxrwxrwx 1 root root   31 Apr 20 19:24 ..data -> ..2021_04_20_19_24_41.755804101
lrwxrwxrwx 1 root root   13 Apr 20 19:24 ca.crt -> ..data/ca.crt
lrwxrwxrwx 1 root root   16 Apr 20 19:24 namespace -> ..data/namespace
lrwxrwxrwx 1 root root   12 Apr 20 19:24 token -> ..data/token
```
And from the Pod:
```
root@mypod:/# ls -la /var/run/secrets/kubernetes.io/serviceaccount/
total 4
drwxrwxrwt 3 root root  140 Apr 20 19:24 .
drwxr-xr-x 3 root root 4096 Apr 20 19:24 ..
drwxr-xr-x 2 root root  100 Apr 20 19:24 ..2021_04_20_19_24_41.755804101
lrwxrwxrwx 1 root root   31 Apr 20 19:24 ..data -> ..2021_04_20_19_24_41.755804101
lrwxrwxrwx 1 root root   13 Apr 20 19:24 ca.crt -> ..data/ca.crt
lrwxrwxrwx 1 root root   16 Apr 20 19:24 namespace -> ..data/namespace
lrwxrwxrwx 1 root root   12 Apr 20 19:24 token -> ..data/token
```

In that case we can trivially use the atomic writer

**Comment by aramase** [comment]
The feature has been implemented and included in `v0.0.22` release. 

#551 is used to track the providers switching to driver write files.

/close
